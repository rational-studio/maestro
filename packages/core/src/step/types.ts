import { type ZodType } from 'zod';
import { type StateCreator, type StoreApi, type StoreMutatorIdentifier } from 'zustand/vanilla';

import { type CLEANUP_ARRAY_EXECUTED } from './constants';

export type CleanupFn = void | (() => void);

export type CleanupFnArray = CleanupFn[] & { [CLEANUP_ARRAY_EXECUTED]?: boolean };

// Transition hooks may be synchronous or asynchronous and return an optional cleanup function.
// Async hooks should resolve to a cleanup function; rejection is handled by the workflow runtime.
export type TransitionHook = () => CleanupFn | Promise<CleanupFn>;

export type DependencyList = readonly unknown[];

export type BuildArgs<Input, Output, Config, Store> = {
  name: string;
  transitionIn: (hook: TransitionHook) => void;
  transitionOut: (hook: TransitionHook) => void;
  effect: (fn: () => CleanupFn, deps?: DependencyList) => void;
  input: Input;
  next: (output: Output) => void;
} & ([Store] extends [undefined] ? {} : { store: Store }) &
  ([Config] extends [undefined] ? {} : { config: Config });

export type StepDef<
  Kind extends string,
  Input,
  Output,
  Config,
  Store,
  Mos extends [StoreMutatorIdentifier, unknown][] = [],
> = {
  kind: Kind;
  inputSchema?: ZodType<Input>;
  outputSchema?: ZodType<Output>;
  configSchema?: ZodType<Config>;
  createStore?: StateCreator<Store, [], Mos>;
};

export type StepAPI = Record<string, unknown>;

export type StepInstance<Input, Output, Config, Api extends StepAPI, Store> = {
  kind: string;
  name: string;
  id: string;
  inputSchema?: ZodType<Input>;
  outputSchema?: ZodType<Output>;
  configSchema?: ZodType<Config>;
  storeApi?: StoreApi<Store>;
  config?: Config;
  build: (args: BuildArgs<Input, Output, Config, Store>) => Api;
};

export type StepCreatorAny = {
  kind: string;
} & ((...args: any[]) => StepInstance<any, any, any, any, any>);

export interface StepCreatorNoConfig<K extends string, Input, Output, Api extends StepAPI, Store> {
  kind: K;
  (name?: string): StepInstance<Input, Output, undefined, Api, Store>;
}

export interface StepCreatorConfig<K extends string, Input, Output, Config, Api extends StepAPI, Store> {
  kind: K;
  (name: string, config: Config): StepInstance<Input, Output, Config, Api, Store>;
  (config: Config): StepInstance<Input, Output, Config, Api, Store>;
}
