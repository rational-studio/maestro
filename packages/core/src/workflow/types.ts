import { type Edge } from '../edge/type';
import { type StepAPI, type StepCreatorAny, type StepInstance } from '../step/types';

export type TransitionStatus = 'notStarted' | 'transitionIn' | 'ready' | 'transitionOut';

export type CurrentStepNotStarted = { status: 'notStarted' };
export type CurrentStepStarted<Creators extends readonly StepCreatorAny[]> = {
  status: Exclude<TransitionStatus, 'notStarted'>;
} & {
  [K in Creators[number]['kind']]: {
    kind: K;
    name: string;
    state: ReturnType<
      ReturnType<
        Extract<
          Creators[number],
          {
            kind: K;
          }
        >
      >['build']
    >;
  };
}[Creators[number]['kind']];

export type CurrentStep<Creators extends readonly StepCreatorAny[]> =
  | CurrentStepNotStarted
  | CurrentStepStarted<Creators>;

// Export/Import schema version
export const WORKFLOW_EXPORT_SCHEMA_VERSION = '1.0.0' as const;

// Shared base for exports
export interface WorkflowExportBase {
  schemaVersion: typeof WORKFLOW_EXPORT_SCHEMA_VERSION;
  libraryVersion?: string;
  inventoryKinds: string[];
  nodes: Array<{
    id: string;
    kind: string;
    name: string;
    config?: unknown;
  }>;
  edges: Array<{
    from: string;
    to: string;
    unidirectional: boolean;
  }>;
}

// Basic configuration export structure
export interface WorkflowExportBasic extends WorkflowExportBase {
  format: 'motif-ts/basic';
}

// Full export extends base with runtime state
export interface WorkflowExportFull extends WorkflowExportBase {
  format: 'motif-ts/full';
  state: {
    current: {
      nodeId?: string | null;
      status: TransitionStatus;
      input?: unknown;
    };
    history: Array<{
      nodeId: string;
      input?: unknown;
    }>;
    stores: Record<string, unknown>;
  };
}

export type WorkflowExport = WorkflowExportBasic | WorkflowExportFull;

export interface WorkflowAPI<Creators extends readonly StepCreatorAny[]> {
  /**
   * Register steps to the workflow.
   * @param nodesArg The steps to register.
   */
  register(nodesArg: ReturnType<Creators[number]> | readonly ReturnType<Creators[number]>[]): void;
  /**
   * Connect two steps together.
   * @param from The step to transition out from.
   * @param to The step to transition into.
   * @param unidirectional Whether the connection is bidirectional or unidirectional. Defaults to bidirectional.
   */
  connect<Input, Output extends Input>(
    from: StepInstance<any, Output, any, any, any>,
    to: StepInstance<Input, any, any, any, any>,
    unidirectional?: boolean,
  ): void;
  /**
   * Connect two steps together using an edge.
   * @param edge The edge to connect.
   */
  connect<Input, Output>(edge: Edge<Input, Output>): void;
  /**
   * Start the workflow.
   */
  start<Input, Output, Config, Api extends StepAPI, Store>(node: StepInstance<Input, Output, Config, Api, Store>): void;
  /**
   * Get the current step.
   */
  getCurrentStep(): CurrentStep<Creators>;
  /**
   * Subscribe to the current step.
   * @param callback The callback to call when the current step changes.
   */
  subscribe(callback: (kind: string, name: string, status: TransitionStatus) => void): () => void;
  /**
   * Back to the previous step.
   */
  back(): void;
  /**
   * Export current workflow configuration or full state to JSON structure.
   * @param mode 'basic' for nodes/edges only; 'full' for including runtime state and history.
   */
  exportWorkflow(mode: 'basic' | 'full'): WorkflowExport;
  /**
   * Import workflow configuration or full state from JSON. Operation is atomic; on any error, no changes are applied.
   * @param data The JSON object to import.
   * @param mode Must match the export format: 'basic' or 'full'.
   */
  importWorkflow(data: WorkflowExport, mode: 'basic' | 'full'): void;
}
