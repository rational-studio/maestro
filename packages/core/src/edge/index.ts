import { type StepInstance } from '../step/types';
import { type Edge } from './type';

// BidirectionalEdge: default connector allowing forward and backward transitions
// Use when connectingA, B) without specifying edge type.
export function edge<I, O extends I>(
  from: StepInstance<any, O, any, any, any>,
  to: StepInstance<I, any, any, any, any>,
  unidirectional = false,
): Edge<I, O> {
  return {
    from,
    to,
    unidirectional,
    validateTransition(outputFrom) {
      // Pass-through without transformation; types are compatible (O extends I)
      return { allow: true, nextInput: outputFrom };
    },
  };
}

export function conditionalEdge<I, O extends I>(
  from: StepInstance<any, O, any, any, any>,
  to: StepInstance<I, any, any, any, any>,
  predicate: (outputFrom: O) => boolean,
  unidirectional = false,
): Edge<I, O> {
  return {
    from,
    to,
    unidirectional,
    validateTransition(outputFrom) {
      const ok = predicate(outputFrom);
      return ok ? { allow: true as const, nextInput: outputFrom } : { allow: false as const };
    },
  };
}

export function transformEdge<I, O>(
  from: StepInstance<any, O, any, any, any>,
  to: StepInstance<I, any, any, any, any>,
  transformer: (outputFrom: O) => I,
  unidirectional = false,
): Edge<I, O> {
  return {
    from,
    to,
    unidirectional,
    validateTransition(outputFrom) {
      try {
        const converted = transformer(outputFrom);
        return { allow: true as const, nextInput: converted };
      } catch (e: any) {
        // Handle conversion errors explicitly
        const msg = e?.message ?? String(e);
        throw new Error(`TransformEdge: failed to convert output -> input. Reason: ${msg}`);
      }
    },
  };
}
