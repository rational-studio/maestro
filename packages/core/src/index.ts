export { step } from './step';
export { workflow } from './workflow';
export { edge, conditionalEdge, transformEdge } from './edge';
export { type Edge } from './edge/type';
