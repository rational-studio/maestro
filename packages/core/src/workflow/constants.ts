/**
 * Configuration and constants for the workflow module.
 * Provides shared values used across workflow utilities and logic.
 */
export const LOG_PREFIX = '[motif-ts]';
