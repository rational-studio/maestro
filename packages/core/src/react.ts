import { useSyncExternalStore, useDebugValue } from 'react';

import { type StepCreatorAny } from './step/types';
import { type CurrentStep, type WorkflowAPI } from './workflow/types';

const getCurrentStepServerSide: () => CurrentStep<[]> = () => ({
  status: 'notStarted',
});

export function useWorkflow<const Creators extends readonly StepCreatorAny[]>({
  subscribe,
  getCurrentStep,
}: WorkflowAPI<Creators>) {
  const slice = useSyncExternalStore(subscribe, getCurrentStep, getCurrentStepServerSide);
  useDebugValue(slice);
  return slice;
}
