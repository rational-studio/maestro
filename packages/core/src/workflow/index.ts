import { edge } from '../edge';
import { type Edge } from '../edge/type';
import { CLEANUP_ARRAY_EXECUTED } from '../step/constants';
import {
  type BuildArgs,
  type CleanupFn,
  type CleanupFnArray,
  type DependencyList,
  type StepAPI,
  type StepCreatorAny,
  type StepInstance,
  type TransitionHook,
} from '../step/types';
import { type WorkflowContext } from './context';
import { computeNextEffects, initialEffects, type EffectDef } from './effects';
import { createImportExportHandlers } from './io';
import { type CurrentStep, type CurrentStepStarted, type TransitionStatus, type WorkflowAPI } from './types';
import { handleAsyncError, isPromise, runOutCleanupOnBack, safeInvokeCleanup } from './utils';
import { validateInventory } from './validators';

export function workflow<const Creators extends readonly StepCreatorAny[]>(inventory: Creators): WorkflowAPI<Creators> {
  const inventoryMap: Map<string, StepCreatorAny> = new Map();
  const nodes = new Set<StepInstance<any, any, any, any, any>>();
  const edges: Edge<any, any>[] = [];
  const history: Array<{
    node: StepInstance<any, any, any, any, any>;
    input: unknown;
    // transitionOut cleanups to be executed when user navigates back to this step
    outCleanupOnBack: CleanupFn[];
  }> = [];
  const subscribers = new Set<(kind: string, name: string, status: TransitionStatus) => void>();

  // States
  let currentStep: CurrentStep<Creators> = { status: 'notStarted' };
  let current: StepInstance<any, any, any, any, any> | undefined;
  let currentApi: StepAPI | undefined;
  let context: WorkflowContext | undefined;
  let contextVersionCounter = 0;

  // Validate duplicate kinds with detailed error information
  validateInventory(inventory);

  for (const creator of inventory) {
    inventoryMap.set(creator.kind, creator);
  }

  Object.freeze(inventory);

  const subscribe = (handler: (kind: string, name: string, status: TransitionStatus) => void) => {
    subscribers.add(handler);
    return () => {
      subscribers.delete(handler);
    };
  };

  const notify = (kind: string, name: string, status: TransitionStatus) => {
    for (const cb of subscribers) {
      cb(kind, name, status);
    }
  };

  const runTransitionInOnce = () => {
    if (!context || context.hasRunIn) {
      return;
    }
    context.inCleanups = [];
    const version = context.version;
    for (let i = 0; i < context.inHooks.length; i++) {
      const hook = context.inHooks[i];
      const result = hook();
      if (typeof result === 'function') {
        context.inCleanups.push(result);
      } else if (isPromise<CleanupFn>(result)) {
        result
          .then((cleanup) => {
            if (typeof cleanup !== 'function') {
              return;
            }
            // If context is still current, register; otherwise, invoke immediately
            if (context && context.version === version) {
              context.inCleanups.push(cleanup);
            } else {
              safeInvokeCleanup(cleanup);
            }
          })
          .catch((err) => handleAsyncError(err, 'transitionIn', i));
      }
    }
    context.hasRunIn = true;
  };

  // Exit sequence: run transitionOut hooks (once, before exit), collect their cleanups to run when coming back.  134   const runExitSequence = (): CleanupFn[] => {
  // Also run effect cleanups and transitionIn cleanups immediately.
  const runExitSequence = (): CleanupFnArray => {
    const outCleanupsForBack: CleanupFnArray = [];
    // mark as not yet executed; used to flush late async cleanups
    outCleanupsForBack[CLEANUP_ARRAY_EXECUTED] = false;
    if (current) {
      currentStep = {
        status: 'transitionOut',
        kind: current.kind,
        name: current.name,
        state: currentApi,
      } as CurrentStepStarted<Creators>;
      notify(current.kind, current.name, 'transitionOut');
    }
    if (context) {
      for (let i = 0; i < context.outHooks.length; i++) {
        const hook = context.outHooks[i];
        const result = hook();
        if (typeof result === 'function') {
          outCleanupsForBack.push(result);
        } else if (isPromise<CleanupFn>(result)) {
          result
            .then((cleanup) => {
              if (typeof cleanup !== 'function') {
                return;
              }
              // If back has already executed for this array, invoke immediately; else collect
              const executed = outCleanupsForBack[CLEANUP_ARRAY_EXECUTED] === true;
              if (executed) {
                safeInvokeCleanup(cleanup);
              } else {
                outCleanupsForBack.push(cleanup);
              }
            })
            .catch((err) => handleAsyncError(err, 'transitionOut', i));
        }
      }
      for (const eff of context.effects) {
        if (typeof eff.cleanup === 'function') {
          eff.cleanup();
        }
      }
      for (const cleanup of context.inCleanups) {
        safeInvokeCleanup(cleanup);
      }
      context.storeUnsub?.();
      context = undefined;
    }
    return outCleanupsForBack;
  };

  const rebuildCurrent = () => {
    if (!current) {
      return;
    }
    const node = current;
    const inputForNode = context?.currentInput;

    const inHooks: TransitionHook[] = [];
    const outHooks: TransitionHook[] = [];
    const effectsDefs: EffectDef[] = [];

    const args: BuildArgs<any, any, any, any> = {
      name: node.name,
      transitionIn: (hook) => {
        if (!context?.hasRunIn) {
          inHooks.push(hook);
        }
      },
      transitionOut: (hook) => {
        outHooks.push(hook);
      },
      effect: (fn, deps) => {
        effectsDefs.push({ run: fn, deps });
      },
      input: inputForNode,
      ...(node.configSchema ? { config: node.config } : {}),
      ...(node.storeApi ? { store: node.storeApi.getState() } : {}),
      next: (output) => {
        const validatedOutput = node.outputSchema ? node.outputSchema.parse(output) : undefined;
        const outgoing = edges.filter((e) => e.from === node);
        const selected: Edge<any, any> | undefined = outgoing[0];
        if (!selected) {
          const prevOutCleanups = runExitSequence();
          history.push({ node, input: inputForNode, outCleanupOnBack: prevOutCleanups });
          return;
        }
        const nextNode = selected.to;
        let nextInput = undefined;
        if (nextNode.inputSchema) {
          nextInput = nextNode.inputSchema.parse(validatedOutput);
        }
        const prevOutCleanups = runExitSequence();
        history.push({ node, input: inputForNode, outCleanupOnBack: prevOutCleanups });
        transitionInto(nextNode, nextInput, false, []);
      },
    };
    const api = node.build(args);

    // Update outHooks to latest
    if (!context) {
      context = {
        hasRunIn: true,
        inHooks,
        inCleanups: [],
        outHooks,
        outCleanupOnBack: [],
        effects: [],
        storeUnsub: undefined,
        currentInput: inputForNode,
        version: ++contextVersionCounter,
      };
    } else {
      context.outHooks = outHooks;
    }

    // Effect diffing / rerun
    context.effects = computeNextEffects(context.effects, effectsDefs);

    currentApi = api;
    currentStep = {
      status: 'ready',
      kind: current.kind,
      name: current.name,
      state: currentApi,
    } as CurrentStepStarted<Creators>;
    // Notify ready on store-driven rebuild
    notify(node.kind, node.name, 'ready');
  };

  const transitionInto = <Input, Output, Config, Api extends StepAPI, Store>(
    node: StepInstance<Input, Output, Config, Api, Store>,
    input: Input,
    isBack: boolean,
    backCleanups: CleanupFn[],
  ) => {
    // Do NOT run exit sequence here; it is handled by callers (next/back)
    current = node;
    currentStep = {
      status: 'transitionIn',
      kind: current.kind,
      name: current.name,
      state: currentApi,
    } as CurrentStepStarted<Creators>;
    // Notify transition entering
    notify(node.kind, node.name, 'transitionIn');

    if (isBack) {
      // execute transitionOut cleanups of the step we are returning to
      // even if none are present yet; late async cleanups will flush immediately
      runOutCleanupOnBack(backCleanups);
    }

    const inHooks: TransitionHook[] = [];
    const outHooks: TransitionHook[] = [];
    const effectsDefs: EffectDef[] = [];

    const next = (output: Output) => {
      const validatedOutput = node.outputSchema ? node.outputSchema.parse(output) : undefined;
      const outgoing = edges.filter((e) => e.from === node);
      // If there are no outgoing edges, end the workflow
      if (outgoing.length === 0) {
        runExitSequence();
        throw new Error('No next step');
      }
      // Try each outgoing edge and pick the first that allows transition
      let selectedEdge: Edge<Output, any> | undefined;
      let nextInput = undefined;
      for (const e of outgoing) {
        const res = e.validateTransition(validatedOutput);
        if (res.allow) {
          selectedEdge = e as Edge<Output, any>;
          nextInput = res.nextInput;
          break;
        }
      }
      if (!selectedEdge) {
        runExitSequence();
        throw new Error('Transition blocked by edge condition');
      }
      const nextNode = selectedEdge.to;
      if (nextNode.inputSchema) {
        nextInput = nextNode.inputSchema.parse(nextInput);
      }
      const prevOutCleanups = runExitSequence();
      history.push({ node, input, outCleanupOnBack: prevOutCleanups });
      transitionInto(nextNode, nextInput, false, []);
    };

    const args: any = {
      name: node.name,
      transitionIn: (hook: TransitionHook) => inHooks.push(hook),
      transitionOut: (hook: TransitionHook) => outHooks.push(hook),
      effect: (fn: () => CleanupFn, deps?: DependencyList) => effectsDefs.push({ run: fn, deps }),
      input,
      ...(node.configSchema ? { config: node.config } : {}),
      ...(node.storeApi ? { store: node.storeApi.getState() } : {}),
      next,
    };
    const api = node.build(args);

    // Initialize context for this step
    context = {
      hasRunIn: false,
      inHooks,
      inCleanups: [],
      outHooks,
      outCleanupOnBack: [],
      effects: [],
      storeUnsub: undefined,
      currentInput: input,
      version: ++contextVersionCounter,
    };
    // Execute transitionIn once
    runTransitionInOnce();

    // Run initial effects (first render)
    context.effects = initialEffects(effectsDefs);

    // Subscribe to data layer changes to rebuild on any change

    if (node.storeApi) {
      context.storeUnsub = node.storeApi.subscribe(() => {
        rebuildCurrent();
      });
    }

    currentApi = api;
    currentStep = {
      status: 'ready' as const,
      kind: current.kind,
      name: current.name,
      state: currentApi,
    } as CurrentStepStarted<Creators>;
    notify(node.kind, node.name, 'ready');
  };

  const register = (nodesArg: ReturnType<Creators[number]> | readonly ReturnType<Creators[number]>[]): void => {
    const list = Array.isArray(nodesArg) ? nodesArg : [nodesArg];
    const allowed = Array.from(inventoryMap.keys()).join(', ');
    for (const node of list) {
      if (!inventoryMap.has(node.kind)) {
        throw new Error(
          `Cannot register StepInstance kind '${node.kind}'. Not listed in inventory. Allowed kinds: [${allowed}]`,
        );
      }
      nodes.add(node);
    }
  };

  const connect = <I, O>(
    fromOrEdge: StepInstance<any, O, any, any, any> | Edge<I, O>,
    to?: StepInstance<I, any, any, any, any>,
    unidirectional = false,
  ): void => {
    if (to) {
      const from = fromOrEdge as StepInstance<any, O, any, any, any>;
      if (!nodes.has(from)) {
        throw new Error(
          `Cannot connect from unregistered StepInstance '${from.id}'. Register the instance before connecting.`,
        );
      }
      if (!nodes.has(to)) {
        throw new Error(
          `Cannot connect to unregistered StepInstance '${to.id}'. Register the instance before connecting.`,
        );
      }
      const e = edge<any, any>(from, to, unidirectional);
      edges.push(e);
    } else {
      const e = fromOrEdge as Edge<I, O>;
      if (!nodes.has(e.from)) {
        throw new Error(
          `Cannot connect from unregistered StepInstance '${e.from.id}'. Register the instance before connecting.`,
        );
      }
      if (!nodes.has(e.to)) {
        throw new Error(
          `Cannot connect to unregistered StepInstance '${e.to.id}'. Register the instance before connecting.`,
        );
      }
      edges.push(e);
    }
  };

  const start = <I, O, C, Api extends StepAPI, Store>(node: StepInstance<I, O, C, Api, Store>) => {
    if (!inventoryMap.has(node.kind)) {
      const allowed = Array.from(inventoryMap.keys()).join(', ');
      throw new Error(
        `Cannot start on StepInstance kind '${node.kind}'. Not listed in inventory. Allowed kinds: [${allowed}]`,
      );
    }
    if (!nodes.has(node)) {
      throw new Error(`Cannot start on unregistered StepInstance '${node.id}'. Register the instance before starting.`);
    }
    transitionInto(node, undefined as I, false, []);
  };

  const getCurrentStep = () => currentStep;
  const getCurrentNode = () => current;
  const getContext = () => context;
  const setNotStarted = () => {
    current = undefined;
    currentApi = undefined;
    context = undefined;
    currentStep = { status: 'notStarted' };
  };

  const back = () => {
    const prev = history.pop();
    if (!prev) {
      return;
    }
    const connecting = edges.find((e) => e.from === prev.node && e.to === current);
    // Enforce edge reversibility: find the edge previously used to reach current from prev
    if (connecting && connecting.unidirectional) {
      throw new Error(
        `Back navigation is not allowed: edge from '${prev.node.id}' to '${current?.id}' is unidirectional`,
      );
    }
    // Run exit sequence for the current step (we are leaving it)
    runExitSequence();
    // When backing into prev.node, run its transitionOut cleanup collected at its previous exit
    transitionInto(prev.node, prev.input, true, prev.outCleanupOnBack);
  };

  // wire import/export handlers from separate module
  const { exportWorkflow, importWorkflow } = createImportExportHandlers({
    inventoryMap,
    nodes,
    edges,
    history,
    getCurrentStep,
    getCurrentNode,
    getContext,
    setNotStarted,
    runExitSequence,
    transitionInto,
  });

  return {
    register,
    connect,
    start,
    getCurrentStep,
    subscribe,
    back,
    exportWorkflow,
    importWorkflow,
  };
}
