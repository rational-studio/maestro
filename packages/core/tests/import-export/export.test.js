import { describe, it, expect } from 'vitest';
import z from 'zod';
import { step, workflow } from '../../src';
import { WORKFLOW_EXPORT_SCHEMA_VERSION } from '../../src/workflow/types';

describe('Export workflow - basic and full', () => {
  it('exports basic configuration with nodes and edges', () => {
    const A = step({ kind: 'A', outputSchema: z.number() }, ({ next }) => ({ go: () => next(1) }));
    const B = step({ kind: 'B', inputSchema: z.number() }, () => ({ done: true }));

    const wf = workflow([A, B]);
    const a = A('a');
    const b = B('b');
    wf.register([a, b]);
    wf.connect(a, b);

    wf.start(a);
    const basic = wf.exportWorkflow('basic');
    // Snapshot ensures full structure correctness (schemaVersion, inventoryKinds, nodes, edges, etc.)
    expect(basic).toMatchSnapshot('basic-export');
    // Minimal invariant check to ensure snapshot remains meaningful
    expect(basic.schemaVersion).toBe(WORKFLOW_EXPORT_SCHEMA_VERSION);
  });

  it('exports full state including current, history and stores', () => {
    const StoreStep = step(
      { kind: 'S', outputSchema: z.number(), configSchema: z.object({ v: z.number() }), createStore: () => ({ v: 0 }) },
      ({ config, next }) => ({ run: () => next(config.v) }),
    );
    const Sink = step({ kind: 'T', inputSchema: z.number() }, ({ input }) => ({ value: input }));

    const wf = workflow([StoreStep, Sink]);
    const s = StoreStep('s', { v: 42 });
    const t = Sink('t');
    wf.register([s, t]);
    wf.connect(s, t);

    wf.start(s);
    const full = wf.exportWorkflow('full');
    // Snapshot full export including current step, history, and stores
    expect(full).toMatchSnapshot('full-export');
    expect(full.schemaVersion).toBe(WORKFLOW_EXPORT_SCHEMA_VERSION);
  });

  it('exports when not started (no current node)', () => {
    const A = step({ kind: 'A' }, () => ({ noop: true }));
    const wf = workflow([A]);
    const a = A('a');
    wf.register(a);

    const full = wf.exportWorkflow('full');
    // Snapshot not-started state to guarantee initial export correctness
    expect(full).toMatchSnapshot('full-export-not-started');
    expect(full.schemaVersion).toBe(WORKFLOW_EXPORT_SCHEMA_VERSION);
  });
});