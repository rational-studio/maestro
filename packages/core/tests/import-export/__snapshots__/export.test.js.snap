// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Export workflow - basic and full > exports basic configuration with nodes and edges > basic-export 1`] = `
{
  "$id": "https://motif-ts.dev/schemas/workflow-export.json",
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "edges": [
    {
      "from": "A_a",
      "to": "B_b",
      "unidirectional": false,
    },
  ],
  "format": "motif-ts/basic",
  "inventoryKinds": [
    "A",
    "B",
  ],
  "libraryVersion": undefined,
  "nodes": [
    {
      "config": undefined,
      "id": "A_a",
      "kind": "A",
      "name": "a",
    },
    {
      "config": undefined,
      "id": "B_b",
      "kind": "B",
      "name": "b",
    },
  ],
  "schemaVersion": "1.0.0",
}
`;

exports[`Export workflow - basic and full > exports full state including current, history and stores > full-export 1`] = `
{
  "$id": "https://motif-ts.dev/schemas/workflow-export.json",
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "edges": [
    {
      "from": "S_s",
      "to": "T_t",
      "unidirectional": false,
    },
  ],
  "format": "motif-ts/full",
  "inventoryKinds": [
    "S",
    "T",
  ],
  "libraryVersion": undefined,
  "nodes": [
    {
      "config": {
        "v": 42,
      },
      "id": "S_s",
      "kind": "S",
      "name": "s",
    },
    {
      "config": undefined,
      "id": "T_t",
      "kind": "T",
      "name": "t",
    },
  ],
  "schemaVersion": "1.0.0",
  "state": {
    "current": {
      "input": undefined,
      "nodeId": "S_s",
      "status": "ready",
    },
    "history": [],
    "stores": {
      "S_s": {
        "v": 0,
      },
    },
  },
}
`;

exports[`Export workflow - basic and full > exports when not started (no current node) > full-export-not-started 1`] = `
{
  "$id": "https://motif-ts.dev/schemas/workflow-export.json",
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "edges": [],
  "format": "motif-ts/full",
  "inventoryKinds": [
    "A",
  ],
  "libraryVersion": undefined,
  "nodes": [
    {
      "config": undefined,
      "id": "A_a",
      "kind": "A",
      "name": "a",
    },
  ],
  "schemaVersion": "1.0.0",
  "state": {
    "current": {
      "input": undefined,
      "nodeId": null,
      "status": "notStarted",
    },
    "history": [],
    "stores": {},
  },
}
`;
